File: openinstall-ios/TrackingAPI.swift
//
//  TrackingAPI.swift
//  OpenInstall iOS SDK
//
//  追踪 API 服务
//

import Foundation

class TrackingAPI {
    
    static let shared = TrackingAPI()
    
    private let baseURL: String
    
    init(baseURL: String = "http://localhost:8080/api") {
        self.baseURL = baseURL
    }
    
    /// 获取安装参数
    /// - Parameter completion: 回调函数，返回参数字典或 nil
    func getInstallParams(completion: @escaping ([String: String]?) -> Void) {
        let fingerprint = DeviceFingerprint.collect()
        let fingerprintId = DeviceFingerprint.generateId(from: fingerprint)
        
        guard let url = URL(string: "\(baseURL)/tracking/get") else {
            completion(nil)
            return
        }
        
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        
        let body: [String: Any] = [
            "fingerprintId": fingerprintId,
            "fingerprint": fingerprint
        ]
        
        guard let jsonData = try? JSONSerialization.data(withJSONObject: body) else {
            completion(nil)
            return
        }
        
        request.httpBody = jsonData
        
        URLSession.shared.dataTask(with: request) { data, response, error in
            guard let data = data,
                  error == nil,
                  let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
                  let success = json["success"] as? Bool,
                  success == true,
                  let dataDict = json["data"] as? [String: Any],
                  let matched = dataDict["matched"] as? Bool,
                  matched == true,
                  let params = dataDict["params"] as? [String: String] else {
                DispatchQueue.main.async {
                    completion(nil)
                }
                return
            }
            
            DispatchQueue.main.async {
                completion(params)
            }
        }.resume()
    }
    
    /// 保存追踪数据（可选，用于调试）
    func saveTrackingData(params: [String: String], completion: @escaping (Bool) -> Void) {
        let fingerprint = DeviceFingerprint.collect()
        let fingerprintId = DeviceFingerprint.generateId(from: fingerprint)
        
        guard let url = URL(string: "\(baseURL)/tracking/save") else {
            completion(false)
            return
        }
        
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        
        let body: [String: Any] = [
            "fingerprintId": fingerprintId,
            "fingerprint": fingerprint,
            "params": params,
            "timestamp": Int(Date().timeIntervalSince1970 * 1000)
        ]
        
        guard let jsonData = try? JSONSerialization.data(withJSONObject: body) else {
            completion(false)
            return
        }
        
        request.httpBody = jsonData
        
        URLSession.shared.dataTask(with: request) { data, response, error in
            guard let data = data,
                  error == nil,
                  let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
                  let success = json["success"] as? Bool else {
                DispatchQueue.main.async {
                    completion(false)
                }
                return
            }
            
            DispatchQueue.main.async {
                completion(success)
            }
        }.resume()
    }
}

File: openinstall-android/TrackingAPI.kt
package com.openinstall.tracking

import android.content.Context
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONObject
import java.util.concurrent.TimeUnit

/**
 * 追踪 API 服务
 */
object TrackingAPI {
    
    private const val BASE_URL = "http://localhost:8080/api"
    
    private val client = OkHttpClient.Builder()
        .connectTimeout(10, TimeUnit.SECONDS)
        .readTimeout(10, TimeUnit.SECONDS)
        .writeTimeout(10, TimeUnit.SECONDS)
        .build()
    
    /**
     * 获取安装参数
     * @param context 上下文
     * @return 参数映射，如果匹配失败返回 null
     */
    suspend fun getInstallParams(context: Context): Map<String, String>? = withContext(Dispatchers.IO) {
        try {
            val fingerprint = DeviceFingerprint.collect(context)
            val fingerprintId = DeviceFingerprint.generateId(fingerprint)
            
            val requestBody = JSONObject().apply {
                put("fingerprintId", fingerprintId)
                put("fingerprint", fingerprint)
            }
            
            val request = Request.Builder()
                .url("$BASE_URL/tracking/get")
                .post(requestBody.toString().toRequestBody("application/json".toMediaType()))
                .addHeader("Content-Type", "application/json")
                .build()
            
            val response = client.newCall(request).execute()
            val responseBody = response.body?.string() ?: return@withContext null
            
            val json = JSONObject(responseBody)
            
            if (json.optBoolean("success", false)) {
                val data = json.optJSONObject("data")
                if (data != null && data.optBoolean("matched", false)) {
                    val params = data.optJSONObject("params")
                    if (params != null) {
                        val result = mutableMapOf<String, String>()
                        val keys = params.keys()
                        while (keys.hasNext()) {
                            val key = keys.next()
                            result[key] = params.optString(key, "")
                        }
                        return@withContext result
                    }
                }
            }
            
            null
        } catch (e: Exception) {
            e.printStackTrace()
            null
        }
    }
    
    /**
     * 保存追踪数据（可选，用于调试）
     */
    suspend fun saveTrackingData(
        context: Context,
        params: Map<String, String>
    ): Boolean = withContext(Dispatchers.IO) {
        try {
            val fingerprint = DeviceFingerprint.collect(context)
            val fingerprintId = DeviceFingerprint.generateId(fingerprint)
            
            val requestBody = JSONObject().apply {
                put("fingerprintId", fingerprintId)
                put("fingerprint", fingerprint)
                put("params", JSONObject(params))
                put("timestamp", System.currentTimeMillis())
            }
            
            val request = Request.Builder()
                .url("$BASE_URL/tracking/save")
                .post(requestBody.toString().toRequestBody("application/json".toMediaType()))
                .addHeader("Content-Type", "application/json")
                .build()
            
            val response = client.newCall(request).execute()
            val responseBody = response.body?.string() ?: return@withContext false
            
            val json = JSONObject(responseBody)
            json.optBoolean("success", false)
        } catch (e: Exception) {
            e.printStackTrace()
            false
        }
    }
}



